src/popup/components/InputAmount.tsx

```diff
@@ -44,67 +44,194 @@ export const InputAmount = ({
   min = 0,
   max,
   readOnly,
+  controls = false,
 }: Props) => {
+  const { assetScale } = walletAddress;
+  const step = 1 / 10 ** assetScale;
   const currencySymbol = getCurrencySymbol(walletAddress.assetCode);
 
-  const validateAmountOnChange = useThrottle(
-    (ev: React.ChangeEvent<HTMLInputElement>) => {
-      const input = ev.target;
-      const value = input.value;
+  const inputRef = React.useRef<HTMLInputElement>(null);
+
+  const formatAmount = React.useCallback(
+    (value: number) => formatNumber(value, assetScale),
+    [assetScale],
+  );
+
+  const handleValue = React.useCallback(
+    (value: string, skipSetValue = false) => {
       const error = validateAmount(value, walletAddress, min, max);
       if (error) {
         onError(error);
       } else {
-        const amountValue = formatNumber(+value, walletAddress.assetScale);
-        onChange(amountValue, input);
+        const input = inputRef.current!;
+        const formattedValue = formatAmount(+value);
+        if (!skipSetValue) {
+          input.value = formattedValue;
+        }
+        onChange(formattedValue, input);
       }
     },
+    [walletAddress, onChange, onError, formatAmount, min, max],
+  );
+
+  const validateAmountOnChange = useThrottle(
+    (ev: React.ChangeEvent<HTMLInputElement>) => {
+      handleValue(ev.target.value, true);
+    },
     350,
     { trailing: true },
   );
 
+  const handleArrowKeys = React.useCallback(
+    (ev: React.KeyboardEvent<HTMLInputElement>) => {
+      const key = ev.key;
+      if (
+        key !== 'ArrowUp' &&
+        key !== 'ArrowDown' &&
+        key !== 'PageUp' &&
+        key !== 'PageDown'
+      ) {
+        return;
+      }
+
+      ev.preventDefault();
+      const input = ev.currentTarget;
+      const isLargeStep = ev.shiftKey || /^Page(Up|Down)$/.test(key);
+      const direction = key === 'ArrowUp' || key === 'PageUp' ? 1 : -1;
+      const amount = isLargeStep ? step * 100 : step;
+      incOrDec(input, direction, amount, formatAmount, handleValue, min, max);
+    },
+    [formatAmount, handleValue, step, min, max],
+  );
+
+  const onKeyDown = React.useCallback(
+    (ev: React.KeyboardEvent<HTMLInputElement>) => {
+      allowOnlyNumericInput(ev);
+      if (!ev.defaultPrevented) {
+        if (readOnly || !controls) return;
+        handleArrowKeys(ev);
+      }
+    },
+    [handleArrowKeys, readOnly, controls],
+  );
+
+  const controlInc = React.useCallback(() => {
+    incOrDec(inputRef.current!, 1, step, formatAmount, handleValue, min, max);
+  }, [step, formatAmount, handleValue, min, max]);
+  const controlDec = React.useCallback(() => {
+    incOrDec(inputRef.current!, -1, step, formatAmount, handleValue, min, max);
+  }, [step, formatAmount, handleValue, min, max]);
+
   return (
     <Input
       id={id}
       type="text"
       inputMode="numeric"
+      ref={inputRef}
```

Comment:
Since `type=number` implicitly has the role of `spinbutton`, but we are using `type=text`.
```suggestion
      role="spinbutton"
      ref={inputRef}
```

Reply:
Should add this only when controls is true though. Will do.